shared-libs/transitions/src/transitions/muting.js

```diff
@@ -36,22 +37,141 @@ const isRelevantReport = (doc, info = {}) =>
-          isContact(doc) &&
-          !doc.muted &&
-          // If initial_replication_date is 'unknown' .getTime() will return NaN, which is an
-          // acceptable value to pass to isMutedInLineage (it will mean that it won't match because
-          // there is no possible mute date that is "after" NaN)
-          mutingUtils.isMutedInLineage(doc, new Date(infoDoc.initial_replication_date).getTime()) &&
-          !infoDoc.muting_history);
+const isRelevantContact = (doc, infoDoc = {}) => {
+  return Boolean(
+    doc &&
+    isContact(doc) &&
+    (isNewContactWithMutedParent(doc, infoDoc) || mutingUtils.isLastUpdatedOffline(doc))
+  );
+};
+
+const processContact = (change) => {
+  let muted;
+  if (isNewContactWithMutedParent(change.doc, change.info)) {
+    muted = new Date();
+  } else {
+    muted = change.doc.muted ? new Date() : false;
+  }
+
+  return mutingUtils
+    .updateRegistrations(utils.getSubjectIds(change.doc), muted)
+    .then(() => mutingUtils.updateMutingHistory(
+      change.doc,
+      new Date(change.info.initial_replication_date).getTime(),
+      muted
+    ))
+    .then(() => {
+      mutingUtils.updateContact(change.doc, muted);
+      return true;
+    });
+};
+
+/**
+ * Given a list of report ids to process
+ * - hydrates the docs
+ * - excludes irrelevant reports
+ * - reads infodocs
+ * - (re)runs muting transition over every report, in sequence, even if it had already ran
+ * - reports should be processed in the same order we have received them
+ * The purpose of this action is to reconcile offline muting events that have already been processed offline, but due
+ * to characteristics of CouchDB + PouchDB sync + changes watching, there is no guarantee that we process these
+ * changes in their chronological order naturally.
+ * The reportIds parameter is a list of reports that have been created offline _after_ the currently processed report,
+ * have have affected one contact that this report has updated.
+ * This resolves most "conflicts" but does not guarantee a consistent muting state for every case.
+ *
+ * @param {Array<string>} reportIds - an ordered list of report uuids to be processed
+ * @return {Promise}
+ */
+const replayOfflineMutingEvents = (reportIds = []) => {
+  if (!reportIds.length) {
+    return Promise.resolve();
+  }
+
+  return mutingUtils.lineage
+    .fetchHydratedDocs(reportIds)
+    .then(hydratedReports => {
+      hydratedReports = hydratedReports.filter(doc => isRelevantReport(doc, {}));
+
+      const changes = hydratedReports.map(report => ({ id: report._id }));
+      return mutingUtils.infodoc.bulkGet(changes).then(infoDocs => {
+        let promiseChain = Promise.resolve();
+        reportIds.forEach(reportId => {
+          const hydratedReport = hydratedReports.find(report => report._id === reportId);
+          if (!hydratedReport) {
+            return;
+          }
+          promiseChain = promiseChain.then(() => runTransition(hydratedReport, infoDocs));
```

Comment:
I think this is equivalent and easier to read...

```suggestion
          if (hydratedReport) {
            promiseChain = promiseChain.then(() => runTransition(hydratedReport, infoDocs));
          }
```